docs/enhancements/dual-stack-vips.md

```diff
@@ -0,0 +1,104 @@
+
+The current implementation of dual-stack support assumes that everything in the cluster should be reachable via IPv4 and that IPv6 is an addition for the workload running in the cluster. This assumption makes it impossible to interact with such a cluster using IPv6-only client, i.e. makes it impossible for IPv6-only Hub to manage a dual-stack Spoke.
+
+### Goals
+
+* Allow configuring both IPv4 and IPv6 VIP for dual-stack clusters.
+
+### Non-goals
+
+* Allow configuring more than single IPv4 and single IPv6 VIP. This is still not supported by the OCP.
+
+## Proposal
+
+### User Stories
+
+* As an operator of the IPv6-only Hub cluster I want to deploy dual-stack Spoke cluster.
+
+* As a deployer of a dual stack OCP cluster I want to run IPv6-only applications that need access to the API and Ingress VIPs.
+
+### API Extensions
+
+New fields will have to be added to the Cluster object to persist the specified VIPs.
+
+### Implementation Details
+
+#### Phase 1
+
+`cluster-create-params`, `v2-cluster-update-params` and `cluster` will get new fields
+
+* `api_vips` of type `array`
+* `ingress_vips` of type `array`
+
+Respective AgentClusterInstall CR will also get those new fields.
+
+The database schema will be updated by creating new tables `api_vips` and `ingress_vips` with the following columns
+
+* `cluster_id`
+* `vip`
+
+#### Phase 2
+
+A set of functions for cluster creation and cluster update will provide a compatiblity between already existing fields (`api_vip` and `ingress_vip`) and new ones (`api_vips` and `ingress_vips`). The behaviour will be as follows
+
+* If only `api_vip` is provided, the `api_vips` will be automatically populated by a simple conversion from `string` to `[]string`. This guarantees no change in behaviour for users not aware of introduction of the new field.
+
+* If only `api_vips` are provided, the `api_vip` will be populated by taking the first value from `[]string`. This guarantees that older versions of the UI do not break when interacting with the new field.
+
+* If both `api_vip` and `api_vips` are provided, the value of `api_vips` will take precedence. This is an undesired behaviour indicating use of the wrong syntax or client but for the completeness the behaviour has to be defined.
+
+All the points described above will behave in an equivalent way for `ingress_vip` and `ingress_vips`.
+
+#### Phase 3
+
+After providing a required period of backwards-compatibility the `api_vip` and `ingress_vip` fields will be deprecated and removed. The value in the database will be nulled. Removal of the column in the database will follow.
+
+### Risks and Mitigations
+
+The similar manouever of introducing `[]string` in place of `string` has already happened in the past when implementing the initial dual-stack support. An example of the field that got transformed is `machine_network_cidr` that became `machine_networks`.
+
+There is already an experience and established path, therefore the risk is considered minimal.
```

Comment:
That's true, but we've still never broken a k8s API (the machine networks change happened before).  We'll have to create a new API version for phase 3, but that's fine.

Reply:
You are right. To be honest what I meant here was the fact that in the 1st dual-stack iteration we *broke* API for real by changing the type of an existing field. That's why my focus here was to explain that we will not repeat this mistake.

For the K8s API you are right, we already had `[]string` fields back then. Here we need to introduce a new field, release a new version (keeping both fields aside) and then in the future release yet newer version with the old field removed

Reply:
That sounds like the path to have the least amount of API breakage. Before when the API was changed it was broken but we had the excuse of not being a released product. But since AI has gone GA now, you have to take more steps to change an API like this. Sounds like a good plan.